frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/AuthenticationAuthCodeHandler.java

```diff
@@ -0,0 +1,128 @@
+    private final DynamoService dynamoService;
+    private final AuthorisationCodeService authorisationCodeService;
+
+    public AuthenticationAuthCodeHandler(
+            DynamoAuthCodeService dynamoAuthCodeService,
+            DynamoService dynamoService,
+            ConfigurationService configurationService,
+            AuthorisationCodeService authorisationCodeService) {
+        this.gson = new GsonBuilder().create();
+        this.dynamoAuthCodeService = dynamoAuthCodeService;
+        this.dynamoService = dynamoService;
+        this.configurationService = configurationService;
+        this.authorisationCodeService = authorisationCodeService;
+    }
+
+    public AuthenticationAuthCodeHandler(ConfigurationService configurationService) {
+        this.gson = new GsonBuilder().create();
+        this.configurationService = configurationService;
+        this.dynamoAuthCodeService = new DynamoAuthCodeService(configurationService);
+        this.dynamoService = new DynamoService(configurationService);
+        this.authorisationCodeService = new AuthorisationCodeService(configurationService);
+    }
+
+    @Override
+    public APIGatewayProxyResponseEvent handleRequest(
+            APIGatewayProxyRequestEvent input, Context context) {
+        ThreadContext.clearMap();
+        return segmentedFunctionCall(
+                "auth-code::" + getClass().getSimpleName(),
+                () -> authCodeRequestHandler(input, context));
+    }
+
+    private APIGatewayProxyResponseEvent authCodeRequestHandler(
+            APIGatewayProxyRequestEvent input, Context context) throws Json.JsonException {
+        attachLogFieldToLogs(AWS_REQUEST_ID, context.getAwsRequestId());
+
+        if (configurationService.isAuthCodeHandlerEnabled()) {
+            var authCodeRequest = gson.fromJson(input.getBody(), AuthCodeRequest.class);
+
+            var authenticationAuthCodeError =
+                    validateAuthenticationAuthCodeRequest(authCodeRequest);
+            if (authenticationAuthCodeError.isPresent()) {
+                LOG.info("Error message: {}", authenticationAuthCodeError.get().getMessage());
+                return generateApiGatewayProxyErrorResponse(400, authenticationAuthCodeError.get());
+            }
+
+            var userProfile = dynamoService.getUserProfileFromEmail(authCodeRequest.getEmail());
+            if (userProfile.isEmpty()) {
+                LOG.info(
+                        "Error message: Email from session does not have a user profile required to extract Subject ID");
+                return generateApiGatewayProxyErrorResponse(400, ErrorResponse.ERROR_1054);
+            }
+
+            var authorisationCode = new AuthorizationCode();
+            dynamoAuthCodeService.saveAuthCode(
+                    userProfile.get().getSubjectID(),
+                    authorisationCode.getValue(),
+                    authCodeRequest.getRequestedScopeClaims(),
```

Comment:
It's not wrong to have this as a string imho, but @dbes-gds's earlier PR for the next step in this was storing claims as List<String>. Again, no functional problem, but might be neater to keep it uniform?: https://github.com/alphagov/di-authentication-api/pull/3217/files#diff-86eab8a351a5d00cf8271124a49c05d31378a08c3f7b2ad5f2411efd0dbb6316R15

Reply:
(just for context, because I know I mentioned the scope/claim naming before, the link I just gave you still says scopes, but that was overwritten here in a later commit: https://github.com/alphagov/di-authentication-api/commit/6381f9a2e66cf04c9e6e1968200d0d1d25e34510#diff-86eab8a351a5d00cf8271124a49c05d31378a08c3f7b2ad5f2411efd0dbb6316R15)

Reply:
Same as above, the change will be moved to another PR